Add examples of using `filesystem.h` to the KJ tour and docs. (#1340)
capnproto/capnproto: kjdoc/tour.md

* Symlinks, hard links, listing directories, recursive delete, recursive create parents, recursive copy directory, memory mapping, and unnamed temporary files are all exposed and easy to use.
* Sparse files ("hole punching"), copy-on-write file cloning (`FICLONE`, `FICLONERANGE`), `sendfile()`-based copying, `renameat2()` atomic replacements, and more will automatically be used when available.

> ```c++

Reviewer: This needs some kind of a preface I think before just dumping the code. Something like "Here are API examples for accessing files."

Author: Hmm, the rest of the tour uses a pretty similar format: https://github.com/capnproto/capnproto/blob/363cefdf46d37f399ba705ba9d9c93e7fa74bcbe/kjdoc/tour.md#unit-testing-tips
I can change it if you like, though.

> auto fs = kj::newDiskFilesystem();

Reviewer: This needs a comment clarifying that you are recommended to construct this at the outermost part of your program (i.e. main) & dependency inject the implementation where needed (that way, for example, you could have a test implementation that roots you somewhere or be given a user space network filesystem implementation)

Author: Sure, done.

* Symlinks, hard links, listing directories, recursive delete, recursive create parents, recursive copy directory, memory mapping, and unnamed temporary files are all exposed and easy to use.
* Sparse files ("hole punching"), copy-on-write file cloning (`FICLONE`, `FICLONERANGE`), `sendfile()`-based copying, `renameat2()` atomic replacements, and more will automatically be used when available.

```c++

> // NOTE: in most cases, you should construct the `Filesystem` at the entry point of your program,
> // then pass it in as a parameter wherever it's used. This allows test implementations or user-space
> // filesystems.

Reviewer 2: ```suggestion
// NOTE: in most cases, you should construct the `Filesystem` at the entry
// point of your program, then pass it in as a parameter wherever it's used.
// This allows dependency injection of alternative filesystem implementations
// for testing or other interesting use cases.
```

"Dependency injection" is a common term describing the point here.

I wrapped this at 80 chars since this code will go into a document / web page where width may be more constrained.

kj::Own<kj::Filesystem> fs = kj::newDiskFilesystem();

> kj::Own<const kj::ReadableFile>> file = fs->getCurrent().openFile(kj::Path::parse("README.md"));

Reviewer 2: ```suggestion
kj::Own<const kj::ReadableFile>> file =
    fs->getCurrent().openFile(kj::Path::parse("README.md"));
```

kj::String text = file->readAllText();
```

> For large files, consider using mmap when you want to read the full contents of a file, which uses
> on demand paging as contents are accessed rather than buffering the contents into memory first.

Reviewer 2: ```suggestion
For large files, consider using mmap when you want to read the full contents of a file, which uses on demand paging as contents are accessed rather than buffering the contents into memory first.
```

We actually don't wrap prose in this document.

```c++

> kj::Array<const char> fileBuffer = file->mmap(0, file->stat().size).releaseAsChars();

Reviewer 2: ```suggestion
kj::Array<const kj::byte> fileBuffer = file->mmap(0, file->stat().size);
```

I think it's useful to make clear that `mmap()` returns `Array<const byte>`. Converting to chars is a separate concern and I suspect most mmap users may not actually want to do that.

```

See the API reference (header file) for details.

> See `parseFromDirectory` in `capnp/schema-parser.h` for more examples.

Reviewer 2: This document is about KJ, which is intended to be independent of Cap'n Proto and may actually split into its own project someday, so I'm not sure I like referring to capnp source files like this. Also the examples there are intertwined with SchemaParser. If we need better examples for KJ, we should probably write them directly in `filesystem.h`.

Author: I'll add some more examples in filesystem.h.

Reviewer 2: Sadly, userspace virtual filesystems are not a standard feature of operating systems. Technically on Linux there is FUSE but setting that up is complicated and requires privileges.

So yeah, the in-memory FS stuff is really meant for passing to KJ-based libraries. The most common use case would be unit tests, so that you can test code that does file manipulation without creating real files on disk.
